File: transcription_service/src/webserver/features/telemetry/telemetry_keys.py
"""
Key layout and heartbeat timings for the fleet telemetry backplane (B1.7)

Restated by hand from `infra/scribear-redis/src/telemetry/telemetry-keys.ts`
and `telemetry-timing.ts`, which are the contract. This service shares no
package with the Node apps, so the choice is between restating the constants
here and inventing a second spelling of them; the constants are few, and the
reader that must agree with them - admin-server - imports the originals. The
same duplication exists in the other direction: the TypeScript side restates
the worker and provider shapes this service produces.

The pattern is snapshot-plus-index. This host rewrites one snapshot key every
heartbeat under a TTL of several heartbeats, and adds itself to a sorted index
scored by its own wall clock. Nothing ever deletes: liveness is expiry, so a
host that stops writing stops existing and a `kill -9`'d one needs no cleanup
path. The one sharp edge is that sorted-set members carry no TTL of their own,
which is why every publisher prunes the index by score - see the publisher.
"""

TELEMETRY_NAMESPACE = "scribe:v1"

# Index of Transcription Service hosts that have published recently. Member is
# the host identity, score is the publish time in epoch milliseconds.
TRANSCRIPTION_HOST_INDEX_KEY = (
    f"{TELEMETRY_NAMESPACE}:transcription-hosts:index"
)

# How often this host republishes its provider health, in seconds.
#
# Slower than the Node Server beat because this payload is per-host rather than
# per-session and changes far less often: providers are configured at boot,
# worker utilization is already a rolling average, and the remote reachability
# probe behind it is cached, so beating faster would republish an unchanged
# answer.
TRANSCRIPTION_HOST_HEARTBEAT_SEC = 5.0

# Expiry on this host's snapshot key, in milliseconds.
#
# Five heartbeats, which tolerates four consecutive missed writes - a GC pause,
# a Redis failover, a slow tick - before a healthy host disappears from the
# fleet view, while still surfacing a genuinely dead one within seconds.
TRANSCRIPTION_HOST_TTL_MS = int(TRANSCRIPTION_HOST_HEARTBEAT_SEC * 1000) * 5


def transcription_host_snapshot_key(transcription_host: str) -> str:
    """
    Builds the snapshot key for one Transcription Service host

    Args:
        transcription_host  - Identity this host publishes under

    Returns:
        Redis key holding this host's entire provider health body

    One key per host, holding every provider - not one key per provider. A host
    publishes all of its providers from a single read of its registry, so
    splitting them buys no independent freshness, and holding them together is
    what makes each read internally consistent: the worker list and the
    providers that ran on those workers are always from the same instant.

    The identity is interpolated, so it must not contain ':' - Config rejects
    that at boot rather than letting a heartbeat forge a key elsewhere in the
    namespace.
    """
    return f"{TELEMETRY_NAMESPACE}:ts:{transcription_host}"


# How often a session's audio-level stats may be republished, in seconds.
#
# Provisional - matches node-server's per-session heartbeat cadence
# (NODE_HEARTBEAT_MS) as a starting point, not independently validated for
# this payload. There is no existing budget in this repo for how often an
# audio-level readout needs to refresh; revisit if write volume (this is a
# per-session key, not per-host) or staleness becomes a real problem.
AUDIO_STATS_MIN_PUBLISH_INTERVAL_SEC = 2.0

# Expiry on a session's audio-stats key, in milliseconds.
#
# Same "five heartbeats" rule as TRANSCRIPTION_HOST_TTL_MS: tolerates four
# consecutive missed publishes before a live session's audio stats age out of
# the dashboard, while a session that actually closed disappears within
# seconds.
AUDIO_STATS_TTL_MS = int(AUDIO_STATS_MIN_PUBLISH_INTERVAL_SEC * 1000) * 5

# Index of sessions whose audio stats have been published recently. Member is
# the session_uid, score is the publish time in epoch milliseconds.
TRANSCRIPTION_SESSION_AUDIO_INDEX_KEY = (
    f"{TELEMETRY_NAMESPACE}:transcription-sessions-audio:index"
)


def transcription_session_audio_key(session_uid: str) -> str:
    """
    Builds the snapshot key for one live session's audio-level stats

    Args:
        session_uid - Opaque session identifier the caller supplied

    Returns:
        `scribe:v1:audio:{sessionUid}` - deliberately a distinct key family
        from node-server's `scribe:v1:session:{sessionUid}` (owned by
        node-server, a different payload), even though both are keyed by the
        same session_uid.
    """
    return f"{TELEMETRY_NAMESPACE}:audio:{session_uid}"

File: infra/scribear-redis/src/telemetry/telemetry-keys.ts
/**
 * Key layout for the fleet telemetry backplane (B1.7).
 *
 * Every key lives under one namespace so a Redis shared with anything else
 * stays separable, and so the whole telemetry plane can be dropped with a
 * single `SCAN`/`DEL` sweep. The `v1` segment is a schema version: change the
 * meaning or shape of any payload below in a way an old reader would
 * misinterpret and bump it, rather than editing in place. Publishers and
 * readers then simply stop seeing each other, which is a visible outage, where
 * a silently reinterpreted payload is a wrong dashboard.
 *
 * The pattern throughout is snapshot-plus-index:
 *
 * - A **snapshot key** holds the current state of one thing as JSON and is
 *   rewritten every heartbeat with a TTL of several heartbeats. Liveness is
 *   therefore expiry: nothing has to delete anything when a process dies,
 *   because a process that stopped writing stops existing. This is what makes a
 *   `kill -9`'d instance's rooms leave the fleet view on their own.
 * - An **index** is a sorted set whose members are the identifiers of live
 *   snapshots, scored by the publisher's wall clock in epoch milliseconds. It
 *   exists so a reader can enumerate the fleet with one `ZRANGEBYSCORE` plus
 *   one `MGET` instead of a `SCAN`, which is O(keyspace) and unusable on a hot
 *   path. Sorted sets have no TTL per member, so a reader must range from
 *   `now - TTL` rather than trusting the whole set, and publishers prune with
 *   `ZREMRANGEBYSCORE`.
 *
 * Scores are wall-clock epoch milliseconds from the writing host, which means
 * they carry that host's clock skew. That is fine for the one thing they are
 * used for - deciding whether a snapshot is within a multi-second TTL - and is
 * not fine for anything finer. Latency figures are never computed from these;
 * they are measured inside a single process and travel inside the payload.
 *
 * Identifiers interpolated below (`sessionUid`, `nodeInstanceId`,
 * `transcriptionHost`) must not contain `:`, or they would forge a key in
 * another part of the namespace. Session uids are uuids and hosts are
 * hostnames, so this holds today; the services that supply the configurable
 * ones validate them at config load rather than here, so a bad value fails at
 * boot instead of once per heartbeat.
 */
export const TELEMETRY_NAMESPACE = 'scribe:v1';

/**
 * Index of Node Server instances that have published recently.
 *
 * Member is `nodeInstanceId`, score is the publish time. Read this rather than
 * deriving the instance list from session records: an instance with no
 * sessions is idle, not absent, and only its own snapshot can say so.
 */
export const NODE_INDEX_KEY = `${TELEMETRY_NAMESPACE}:nodes:index`;

/**
 * Process-wide telemetry for one Node Server instance.
 *
 * Value is a `NodeSnapshot`.
 */
export function nodeSnapshotKey(nodeInstanceId: string): string {
  return `${TELEMETRY_NAMESPACE}:node:${nodeInstanceId}`;
}

/**
 * Index of sessions live anywhere in the fleet.
 *
 * Member is `sessionUid`, score is the publish time of its snapshot.
 */
export const SESSION_INDEX_KEY = `${TELEMETRY_NAMESPACE}:sessions:index`;

/**
 * One live session's telemetry, written by the Node Server instance that owns
 * it.
 *
 * Value is a `SessionSnapshot`. Sticky routing guarantees a single owner per
 * session, so there is exactly one writer per key; were that to change,
 * last-writer-wins on the whole value is still coherent, because both owners
 * would be describing the same session.
 */
export function sessionSnapshotKey(sessionUid: string): string {
  return `${TELEMETRY_NAMESPACE}:session:${sessionUid}`;
}

/**
 * Which Node Server instance owns a session's upstream.
 *
 * Value is a bare `nodeInstanceId`, not JSON: it is read to answer "who do I
 * ask about this room", and keeping it a plain string means a reader can use
 * it without parsing. It is a separate key from the session snapshot so that
 * lookup does not have to fetch and parse the whole record.
 */
export function sessionRouteKey(sessionUid: string): string {
  return `${TELEMETRY_NAMESPACE}:route:${sessionUid}`;
}

/**
 * Index of Transcription Service hosts that have published recently.
 *
 * Member is `transcriptionHost`, score is the publish time.
 */
export const TRANSCRIPTION_HOST_INDEX_KEY = `${TELEMETRY_NAMESPACE}:transcription-hosts:index`;

/**
 * One Transcription Service host's provider health, workers and counters.
 *
 * Value is a `TranscriptionHostSnapshot` - the entire `GET /providers/health`
 * body, not one key per provider as the first draft of this schema had it. A
 * host publishes all of its providers from a single read of its registry, so
 * splitting them buys no independent freshness while costing a second index
 * and a composite member that would have to be parsed back apart. Holding them
 * together also makes each read internally consistent: the worker list and the
 * providers that ran on those workers are always from the same instant.
 */
export function transcriptionHostSnapshotKey(
  transcriptionHost: string,
): string {
  return `${TELEMETRY_NAMESPACE}:ts:${transcriptionHost}`;
}
